crates/circuit/src/parameter/parameter_expression.rs

```diff
@@ -0,0 +1,2162 @@
+            }
+            Ok(
+                ParameterExpression::new(&SymbolExpr::Value(Value::from(r)), &HashMap::new())
+                    .into(),
+            )
+        } else if let Ok(parameter) = ob.extract::<PyParameter>() {
+            Ok(ParameterExpression::from_symbol(&parameter.symbol).into())
+        } else if let Ok(element) = ob.extract::<PyParameterVectorElement>() {
+            Ok(ParameterExpression::from_symbol(&element.symbol).into())
+        } else {
+            ob.extract::<PyParameterExpression>()
+        }
+    }
+}
+
+#[pymethods]
+impl PyParameterExpression {
+    /// This is a **strictly internal** constructor and **should not be used**.
+    /// It is subject to arbitrary change in between Qiskit versions and cannot be relied on.
+    /// Parameter expressions should always be constructed from applying operations on
+    /// parameters, or by loading via QPY.
+    ///
+    /// The input values are allowed to be None for pickling purposes.
+    #[new]
+    #[pyo3(signature = (name_map=None, expr=None))]
+    pub fn py_new(
+        name_map: Option<HashMap<String, PyParameter>>,
+        expr: Option<String>,
+    ) -> PyResult<Self> {
+        match (name_map, expr) {
+            (None, None) => Ok(Self::default()),
+            (Some(name_map), Some(expr)) => {
+                // We first parse the expression and then update the symbols with the ones
+                // the user provided. The replacement relies on the names to match.
+                // This is hacky and we likely want a more reliably conversion from a SymPy object,
+                // if we decide we want to continue supporting this.
+                let expr = parse_expression(&expr)
+                    .map_err(|_| PyRuntimeError::new_err("Failed parsing input expression"))?;
+                let symbol_map: HashMap<String, Symbol> = name_map
+                    .iter()
+                    .map(|(string, param)| (string.clone(), param.symbol.clone()))
+                    .collect();
+
+                let replaced_expr = replace_symbol(&expr, &symbol_map);
+
+                let inner = ParameterExpression::new(&replaced_expr, &symbol_map);
+                Ok(Self {
+                    inner: Arc::new(RwLock::new(inner)),
+                    name_map: name_map,
+                })
+            }
+            _ => Err(PyValueError::new_err(
+                "Pass either both a name_map and expr, or neither",
+            )),
+        }
+    }
+
+    /// TODO Check whether this is required, this was not part of the parameters API before.
+    /// Marked as private for now.
+    #[allow(non_snake_case)]
+    #[staticmethod]
+    pub fn _Value(value: &Bound<PyAny>) -> PyResult<Self> {
+        Self::extract_coerce(value)
+    }
+
+    /// Check if the expression corresponds to a plain symbol.
+    ///
+    /// TODO can we delete this? Not part of public interface before.
```

Comment:
If we don't, then it will be. Just making a note here to delete or prefix with `_`.